packages/cli/src/parallel-tests/run-all-tests.ts

```diff
@@ -0,0 +1,248 @@
+  });
+
+  const testRunUrl = getTestRunUrl(testRun);
+  onTestRunCreated?.({
+    id: testRun.id,
+    url: testRunUrl,
+    status: "Running",
+    progress: {
+      failedTestCases: 0,
+      passedTestCases: 0,
+      runningTestCases: testCases.length,
+    },
+  });
+  logger.info("");
+  logger.info(`Test run URL: ${testRunUrl}`);
+  logger.info("");
+
+  const getResults = async () => {
+    if (parallelTasks == null || parallelTasks > 1) {
+      const results = await runAllTestsInParallel({
+        config,
+        client,
+        testRun,
+        executionOptions,
+        screenshottingOptions,
+        apiToken,
+        commitSha,
+        appUrl,
+        useAssetsSnapshottedInBaseSimulation,
+        parallelTasks,
+        deflake,
+        cachedTestRunResults,
+        replayEventsDependencies,
+        baseCommitSha,
+        onTestFinished: (progress) => {
+          onTestFinished?.({
+            id: testRun.id,
+            url: testRunUrl,
+            status: "Running",
+            progress,
+          });
+        },
+      });
+      return results;
+    }
+
+    const results: TestCaseResult[] = [...cachedTestRunResults];
+    const testsToRun = await getTestsToRun({
+      testCases,
+      cachedTestRunResults,
+      client,
+      baseCommitSha: baseCommitSha ?? null,
+    });
+    for (const testCase of testsToRun) {
+      const result = await deflakeReplayCommandHandler({
+        replayTarget: getReplayTargetForTestCase({
+          useAssetsSnapshottedInBaseSimulation,
+          appUrl,
+          testCase,
+        }),
+        executionOptions,
+        screenshottingOptions,
+        testCase,
+        apiToken,
+        commitSha,
+        deflake,
+        generatedBy: { type: "testRun", runId: testRun.id },
+        testRunId: testRun.id,
+        replayEventsDependencies,
+      });
+      results.push(result);
+      onTestFinished?.({
+        id: testRun.id,
+        url: testRunUrl,
+        status: "Running",
+        progress: {
+          ...getResultCounts(results),
+          runningTestCases: testsToRun.length - results.length,
+        },
+      });
+      await putTestRunResults({
+        client,
+        testRunId: testRun.id,
+        status: "Running",
+        resultData: { results },
+      });
+    }
+    return sortResults({ results, testCases });
+  };
+
+  const results = await getResults();
+
+  const runAllFailure = results.find(({ result }) => result === "fail");
+  const overallStatus = runAllFailure ? "Failure" : "Success";
+  await putTestRunResults({
+    client,
+    testRunId: testRun.id,
+    status: overallStatus,
+    resultData: { results },
+  });
+
+  logger.info("");
+  logger.info("Results");
+  logger.info("=======");
+  logger.info(`URL: ${testRunUrl}`);
+  logger.info("=======");
+  results.forEach(({ title, result }) => {
+    logger.info(`${title} => ${result}`);
+  });
+
+  if (githubSummary) {
+    await writeGitHubSummary({ testRunUrl, results });
```

Comment:
I'm planning to pull this out to live in the action in a later PR, and deprecate the githubSummary param.